lib/solid.ex

```diff
@@ -87,6 +88,55 @@ defmodule Solid do
     {Enum.reverse(result), context}
   end
 
+  defp maybe_trim({:text, _data}, result, acc, context) do
+    result = maybe_trim_current(result, context)
+    context = Map.put(context, :trim_next, false)
+    {result, acc, context}
+  end
+
+  defp maybe_trim({:object, data}, result, acc, context) do
+    [trim_previous] = Keyword.get(data, :trim_previous)
+    acc = maybe_trim_previous(acc, trim_previous)
+
+    result = maybe_trim_current(result, context)
+
+    [trim_next] = Keyword.get(data, :trim_next)
+    context = Map.put(context, :trim_next, trim_next)
+
+    {result, acc, context}
+  end
+
+  defp maybe_trim({:tag, _data}, result, acc, context) do
+    context = Map.put(context, :trim_next, false)
+    {result, acc, context}
+  end
+
+  defp maybe_trim_current(result, context) do
+    if context.trim_next do
+      trim_leading(result)
+    else
+      result
+    end
+  end
+
+  defp maybe_trim_previous(acc, flag) do
+    if flag && !Enum.empty?(acc) do
+      [prev | tail] = acc
+      trimmed_prev = trim_trailing(prev)
+      [trimmed_prev | tail]
+    else
+      acc
+    end
+  end
```

Comment:
I wonder if we could simplify maybe_trim_previous by splitting into multiple definitions?

```elixir  
  defp maybe_trim_previous(acc, false), do: acc
  defp maybe_trim_previous([], _), do: acc
  defp maybe_trim_previous([prev | tail], true) do
      trimmed_prev = trim_trailing(prev)
      [trimmed_prev | tail]
  end
```

What do you think?

Reply:
That's a lot better! Thanks for the hint.